### PR TITLE
CJM-31519 Set adobeAepValidationToken URL param in preview requests

### PR DESCRIPTION
**Pull request by XDex**
## Description

We need to set adobeAepValidationToken param in Konductor requests when it is present in page URL in order for Konductor to enable validation (debugging + preview)

## Related Issue

https://jira.corp.adobe.com/browse/CJM-31519

## Motivation and Context

This is required for supporting Preview links, see [Preview spec wiki](https://wiki.corp.adobe.com/x/Y4rymQ)
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.

### CODE REVIEW

**Comment by jonsnyder** [comment]
Why does this need to be a URL parameter and not passed in meta like we do for other things?
Also, why do we need to do this at all? Couldn't Konductor read the original query string parameter from the referrer header?

**Comment by ninaceban** [comment]
> Why does this need to be a URL parameter and not passed in meta like we do for other things? Also, why do we need to do this at all? Couldn't Konductor read the original query string parameter from the referrer header?

Konductor looks at the `adobeAepValidationToken` query string param or header right now. We can't use header because it will trigger a HTTP Options request.
I agree with Jon here, if Konductor can read the `event.meta` the best place for it would be to pass it in the `event.meta`.

**Comment by XDex** [comment]
> I agree with Jon here, if Konductor can read the event.meta the best place for it would be to pass it in the event.meta.

@jonsnyder @ninaceban Right, though I'm not sure Konductor team will be on board with this proposal, since in the current implementation `validationSessionId` is set at http request level based on the presence of a request header/URL param, while with this proposal request body would also need to be inspected, which implies bigger changes in Konductor. cc: @raresvla @flore77

**Comment by XDex** [comment]
@jonsnyder Regarding `Referer` header - I don't think we can rely on it, since it's optional and even when it's present all major browsers are now stripping it of any query params:
"As of March 2021, by default [Chrome](https://en.wikipedia.org/wiki/Google_Chrome),[[4]](https://en.wikipedia.org/wiki/HTTP_referer#cite_note-N9xNj-4) [Chromium](https://en.wikipedia.org/wiki/Chromium_(web_browser))-based [Edge](https://en.wikipedia.org/wiki/Microsoft_Edge), [Firefox](https://en.wikipedia.org/wiki/Firefox),[[5]](https://en.wikipedia.org/wiki/HTTP_referer#cite_note-6l6dr-5) [Safari](https://en.wikipedia.org/wiki/Safari_(web_browser))[[6]](https://en.wikipedia.org/wiki/HTTP_referer#cite_note-6) default to sending only the origin in cross-origin requests, stripping out everything but the domain name."

**Comment by raresvla** [comment]
@XDex, @jonsnyder The standard way of passing this validation token is either through query parameter or header. Since this is embraced by all services that integrate with Assurance, I would not add something else -- especially since the Debugger extension already passes that token to us as a query param.

**Review by jonsnyder — APPROVED** [review]
I'd love to see a functional test with this. I think it would require creating a griffon session though. Currently we don't have any functional tests that require a token. We do in Tags extension. Is it possible to setup a griffon session and preview experience via API? If so I can change the tests to have a token, if you can add the code for the test. (We could do this as a separate ticket if needed)

**Comment by XDex** [comment]
@jonsnyder Thank you for approving the PR! I need another approval in order to be able to merge it, @ninaceban could you please take a look?
Regarding adding a functional test - you're right, we need to be able to create Griffon sessions for that, and this in and of itself is a separate bigger task since we need to store Griffon auth secrets, make the calls for session creation and log retrieval and parse the event response. It should definitely be on the backlog, but I'm not sure we'll get to it any time soon, since it's probably on the lower priority list, considering our current commitments. cc: @jfkhoury @5amfung

**Review by jfkhoury — APPROVED** [review]
Thank you and good work.

@XDex Can you please create a functional test ticket on your side?

**Comment by XDex** [comment]
As per Dustin Long's [latest comment](https://wiki.corp.adobe.com/display/CJM/AJO+Web+Channel+in+Campaign+-+Preview+and+Test+Mode?focusedCommentId=2678628171&src=mail&src.mail.product=confluence-server&src.mail.timestamp=1659578956001&src.mail.notification=com.atlassian.confluence.plugins.confluence-mentions-plugin%3Amention-created-notification&src.mail.recipient=8a85e89054f739c10154fb4edc8d0035&src.mail.action=view#comment-2678628171) on the AJO Preview & Test wiki, I'll be adjusting this PR accordingly, so we shouldn't be merging it yet..

**Comment by XDex** [comment]
@ninaceban @jonsnyder I've implemented the changes based on latest input from the Griffon team, so now the validation token is generated based on the Griffon session id and a client id uuid persisted in localStorage. Please take another look when you get a chance. Thanks!

File: src/utils/request/createGetValidationQuery.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  return clientId;
+};
+
+export default ({ window }) => {
+  const createNamespacedStorage = injectStorage(window);
+  const storage = createNamespacedStorage(VALIDATION_NAMESPACE);
+  return () => {
+    const parsedQuery = queryString.parse(window.location.search);
+    const validationSessionId = parsedQuery[VALIDATION_URL_PARAM];
+    if (!validationSessionId) {
```
Comment (ninaceban) [inline src/utils/request/createGetValidationQuery.js:34]: here as well, we should create the namespaced storage only if `validationSessionId` present.

File: test/unit/specs/utils/request/createGetValidationQuery.spec.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+describe("createGetValidationQuery", () => {
+  it("gets validation query", () => {
+    let result;
+    let token;
+    let firstClientId;
+    let clientId;
+    const getValidationQuery = createGetValidationQuery({
+      window: win,
+      injectStorage
```
Comment (ninaceban) [inline test/unit/specs/utils/request/createGetValidationQuery.spec.js:32]: `createGetValidationQuery` allows only `window` param.

File: src/core/edgeNetwork/injectSendEdgeNetworkRequest.js [hunk 3]
```diff
@@ -21,7 +21,8 @@ export default ({
   sendNetworkRequest,
   createResponse,
   processWarningsAndErrors,
-  getLocationHint
+  getLocationHint,
+  getValidationQuery = () => ""
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/core/edgeNetwork/injectSendEdgeNetworkRequest.js:25]: Right, fixing

File: src/core/edgeNetwork/injectSendEdgeNetworkRequest.js [hunk 4]
```diff
@@ -56,7 +57,7 @@ export default ({
         const edgeBasePathWithLocationHint = locationHint
           ? `${edgeBasePath}/${locationHint}`
           : edgeBasePath;
-        const url = `https://${endpointDomain}/${edgeBasePathWithLocationHint}/${apiVersion}/${request.getAction()}?configId=${edgeConfigId}&requestId=${request.getId()}`;
+        const url = `https://${endpointDomain}/${edgeBasePathWithLocationHint}/${apiVersion}/${request.getAction()}?configId=${edgeConfigId}&requestId=${request.getId()}${getValidationQuery()}`;
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/core/edgeNetwork/injectSendEdgeNetworkRequest.js:60]: Sure, renaming

File: src/core/index.js [hunk 5]
```diff
@@ -66,6 +67,7 @@ const sendFetchRequest = isFunction(fetch)
   ? injectSendFetchRequest({ fetch })
   : injectSendXhrRequest({ XMLHttpRequest });
 const fireReferrerHideableImage = injectFireReferrerHideableImage();
+const getValidationQuery = createGetValidationQuery({ window });
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/core/index.js:70]: Right, fixing

File: src/utils/request/createGetValidationQuery.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+governing permissions and limitations under the License.
+*/
+
+import { injectStorage, uuid, queryString } from "../index";
+
+const VALIDATION_URL_PARAM = "adb_validation_sessionid";
+const VALIDATION_NAMESPACE = "validation.";
+const CLIENT_ID = "clientId";
+
+const getValidationClientId = storage => {
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/utils/request/createGetValidationQuery.js:19]: Right, most accurately it should be `getOrCreateAssuranceClientId`

File: src/utils/request/createGetValidationQuery.js (hunk 1 above)
Comment (XDex) [reply src/utils/request/createGetValidationQuery.js:34]: No, we actually want to create the storage only once when `createGetAssuranceValidationTokenParams` is called, and not each time `getAssuranceValidationTokenParams()` is called.
Also, namespaced storage is not actually created in localStorage until `setItem()` is called, so it's not a problem even when we don't have a preview param in the URL.

File: test/unit/specs/utils/request/createGetValidationQuery.spec.js (hunk 2 above)
Comment (XDex) [reply test/unit/specs/utils/request/createGetValidationQuery.spec.js:32]: Fixed, thx

File: src/utils/request/createGetValidationQuery.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { injectStorage, uuid, queryString } from "../index";
+
+const VALIDATION_URL_PARAM = "adb_validation_sessionid";
+const VALIDATION_NAMESPACE = "validation.";
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/utils/request/createGetValidationQuery.js:16]: AFAIK the term "Validation" refers to both preview and debugging in Griffon/Assurance, so not sure what the problem is, but let me change it to `ASSURANCE_VALIDATION_NAMESPACE`

**Comment by XDex** [comment]
@ninaceban I've addressed all review comments, please take another look. Thanks!
> We need functional tests where we assert:
>    that the client ID is saved in the storage
>    that we retrieve and add to the token the client ID from the storage

These assertions are already fully covered in the existing unit test that uses actual browser's `localStorage` when the test runs, so there's no need for a functional test.

> Question: do we need to clear it from the storage? Can multiple assurance sessions have same clientId?

No, since this represents the unique Griffon validation client id, we need to keep it for future preview and/or debugging sessions either in alloy or AEPD (that would also be looking for it in `localStorage`)

**Comment by jonsnyder** [comment]
Closing in favor of #921

**Pull request closed without merging**